tests/epyccel/recognised_functions/test_numpy_types.py

```diff
@@ -0,0 +1,1234 @@
+    s = shape(a)
+    return len(s), s[0], s[1], a[0,0], a[1,0]
+
+@types('bool[:,:]')
+@types('int[:,:]')
+@types('int8[:,:]')
+@types('int16[:,:]')
+@types('int32[:,:]')
+@types('int64[:,:]')
+@types('float[:,:]')
+@types('float32[:,:]')
+@types('float64[:,:]')
+def get_int32_arr_2d(arr):
+    from numpy import int32, shape
+    a = int32(arr)
+    s = shape(a)
+    return len(s), s[0], s[1], a[0,0], a[1,0]
+
+@types('bool[:,:]')
+@types('int[:,:]')
+@types('int8[:,:]')
+@types('int16[:,:]')
+@types('int32[:,:]')
+@types('int64[:,:]')
+@types('float[:,:]')
+@types('float32[:,:]')
+@types('float64[:,:]')
+def get_int16_arr_2d(arr):
+    from numpy import int16, shape
+    a = int16(arr)
+    s = shape(a)
+    return len(s), s[0], s[1], a[0,0], a[1,0]
+
+@types('bool[:,:]')
+@types('int[:,:]')
+@types('int8[:,:]')
+@types('int16[:,:]')
+@types('int32[:,:]')
+@types('int64[:,:]')
+@types('float[:,:]')
+@types('float32[:,:]')
+@types('float64[:,:]')
+def get_int8_arr_2d(arr):
+    from numpy import int8, shape
+    a = int8(arr)
+    s = shape(a)
+    return len(s), s[0], s[1], a[0,0], a[1,0]
+
+@pytest.mark.parametrize( 'language', (
+        pytest.param("fortran", marks = [pytest.mark.fortran]),
+        pytest.param("c", marks = [
+            pytest.mark.skip(reason="Arrays not handled yet."),
+            pytest.mark.c]
+        ),
+        pytest.param("python", marks = [
+            pytest.mark.python]
+        )
+    )
+)
+
+@pytest.mark.parametrize( 'get_int', [get_int64_arr_2d, get_int32_arr_2d, get_int16_arr_2d, get_int8_arr_2d])
+def test_numpy_int_array_like_2d(language, get_int):
+
+    size = (2, 5)
+
+    bl = randint(0, 1, size=size, dtype= bool)
+
+    integer8 = randint(min_int8, max_int8, size=size, dtype=np.int8)
+    integer16 = randint(min_int16, max_int16, size=size, dtype=np.int16)
+    integer = randint(min_int, max_int, size=size, dtype=np.int)
+    integer32 = randint(min_int32, max_int32, size=size, dtype=np.int32)
+    integer64 = randint(min_int64, max_int64, size=size, dtype=np.int64)
+
+    fl = uniform(min_float / 2, max_float / 2, size = size)
+    fl32 = uniform(min_float32 / 2, max_float32 / 2, size = size)
+    fl32 = np.float32(fl32)
+    fl64 = uniform(min_float64 / 2, max_float64 / 2, size = size)
+
+    epyccel_func = epyccel(get_int, language=language)
+
+    assert epyccel_func(bl) == get_int(bl)
+    assert epyccel_func(integer8) == get_int(integer8)
+    assert epyccel_func(integer16) == get_int(integer16)
+    assert epyccel_func(integer) == get_int(integer)
+    assert epyccel_func(integer32) == get_int(integer32)
+    # the if block should be removed after resolving (https://github.com/pyccel/pyccel/issues/735).
+    if sys.platform != 'win32':
+        assert epyccel_func(integer64) == get_int(integer64)
+        assert epyccel_func(fl) == get_int(fl)
+        assert epyccel_func(fl64) == get_int(fl64)
+        # Python returns always -32768 wich is less than min_int8, epyccel function returns 0.
```

Comment:
Is this due to the float boundaries? Is this also the case if `fl32=2` for example?

Reply:
i dont think it's due to the float boundaries since it works fine for 1d array tests

Reply:
It works for 1d but not for 2d? I don't see how this could be a boundary problem? Are you sure it works for 1d?
Can you try setting the boundaries to something that would also give an output in a valid range?

Reply:
it works with max/min_int16 and less than that.
Python gives `(2, 2, 5, -32768, -32768)` at overflowing, Fortran gives `(2, 2, 5, 0,  0)` or random values with max/min_int32.
i suggest to fix it by:
```python
    fl32 = uniform(min_float32 / 2, min_float32 / 2, size = size)
    if language == 'fortran' and get_int == get_int16_arr_2d: # pylint: disable=comparison-with-callable
        fl32 = uniform(min_int16, max_int16, size = size)
```

Reply:
I'm not sure it makes sense to test `np.int16(a)` with `a` larger than max/min_int16. It's not super obvious what the result is supposed to be. That said we usually follow what python does so I don't know.
Is `min_int16 == 32768`?

Is it working for c?

It's strange that only `np.int16` gives this problem and not `np.int32` (on machines other than windows)

Reply:
we dont test it in c yet.

Reply:
Is min_int16 == 32768?

Reply:
> Fortran gives (2, 2, 5, 0, 0) or random values with max/min_int32.

int32 not int16?

Reply:
My guess would be that fortran is casting while python is taking the closest int16. Does that look right?

Reply:
yes it does, but how casting can lead to 0 ?

Reply:
Casting often involves reading the data as though it already had the correct type. E.g.:
An unsigned int8 uses 8 bits:
16 in binary gives: `0001 0000`
If we want to cast to an unsigned int4 then we will take 4 of those 8 bits (I'm not sure if it's the first 4 or the last 4) so you will have:
`uint4(16) : 0000 = 0`